packages/confirm-dialog/src/vaadin-confirm-dialog.js

```diff
@@ -276,59 +273,22 @@ class ConfirmDialog extends SlotMixin(ElementMixin(ThemePropertyMixin(PolymerEle
     ];
   }
 
-  /** @protected */
-  get slots() {
-    // NOTE: order in which slots are listed matches the template.
-    return {
-      ...super.slots,
-      header: () => {
-        const h3 = document.createElement('h3');
-        this.__defaultHeader = h3;
-        return h3;
-      },
-      '': () => {
-        const div = document.createElement('div');
-        this.__defaultMessage = div;
-        return div;
-      },
-      'cancel-button': () => {
-        const button = document.createElement('vaadin-button');
-        button.setAttribute('theme', this.cancelTheme);
-        button.textContent = this.cancelText;
-        button._isDefaultButton = true;
-        return button;
-      },
-      'reject-button': () => {
-        const button = document.createElement('vaadin-button');
-        button.setAttribute('theme', this.rejectTheme);
-        button.textContent = this.rejectText;
-        button._isDefaultButton = true;
-        return button;
-      },
-      'confirm-button': () => {
-        const button = document.createElement('vaadin-button');
-        button._isDefaultButton = true;
-        return button;
-      },
-    };
-  }
-
   constructor() {
     super();
-    this.__slottedNodes = [];
-    this._observer = new FlattenedNodesObserver(this, (info) => {
-      this.__onDomChange(info.addedNodes);
-    });
```

Comment:
I wanted to point out that we are effectively replacing one `MutationObserver` (plus whatever `FlattenedNodesObserver` does on top) with five of them. Maybe not a big deal in the scope if this component, but we should maybe be more careful to weigh the maintenance improvement here vs. possible performance impacts.

It seems hard to find good information on this, one article (in german) has an interesting Chart that indicates that you need several thousand observers to have an impact (red line is multiple observers, blue line is multiple observes on one observer):
<img src="https://www.peterkroener.de/uploads/2021/chart.png">

Another aspect is memory usage which will likely increase with the multiple observers approach.

Reply:
Good point. I have an idea to make `FlattenedNodesObserver` in the `SlotController` opt-in.
Will create a separate PR and update a couple of components accordingly.

Reply:
BTW, there is a difference between the following calls:

-  `new FlattenedNodesObserver(this, () => ` 
-  `new FlattenedNodesObserver(slot, () => ` 

In the latter case, it's using [`slotchange`](https://github.com/Polymer/polymer/blob/1e8b246d01ea99adba305ea04c45d26da31f68f1/lib/utils/flattened-nodes-observer.js#L148-L149) event, instead of a `MutationObserver`.
And in theory, the event listener approach should be more performant.

Reply:
Ah good to know, then it's less of an issue for `SlotController`.